### PR TITLE
Bug fix/http2 statistics

### PR DESCRIPTION
**Pull request by mpoeter**
### Scope & Purpose

Fixes the request statistics for HTTP2. As mentioned in #11523, for these requests no statistics were gathered at all.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior change can be verified via automatic tests

AFAIK http2 support is new, so I suppose this fixes does not have to be backported?

### Testing & Verification

This change is already covered by existing tests, such as `api-statistics-spec-grey`.

Perhaps this test can be ungreyed now that statistics for all request types should work correctly?

### CODE REVIEW

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10015/

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10077/

File: arangod/GeneralServer/VstCommTask.cpp [hunk 1]
```diff
@@ -252,6 +252,13 @@ bool VstCommTask<T>::processMessage(velocypack::Buffer<uint8_t> buffer,
     }
     stat.SET_REQUEST_TYPE(req->requestType());
 
+    // process OPTIONS requests similar to how it is done for HTTP
+    if (req->requestType() == rest::RequestType::OPTIONS) {
```
Comment (graetzer) [inline arangod/GeneralServer/VstCommTask.cpp:256]: CORS requests are only used for http preflight requests

File: arangod/RestHandler/RestAdminStatisticsHandler.cpp [hunk 2]
```diff
@@ -56,6 +56,9 @@ RestStatus RestAdminStatisticsHandler::execute() {
   }
 
   if (_request->requestPath() == "/_admin/statistics") {
+    if (_request->value("dont-count-this-request") == "true") {
+      _statistics.SET_IGNORE();
+    }
     getStatistics();
```
Comment (graetzer) [inline arangod/RestHandler/RestAdminStatisticsHandler.cpp:62]: Make this maintenance Mode only ?

Comment (mpoeter) [reply arangod/RestHandler/RestAdminStatisticsHandler.cpp:62]: This is required by the integration tests. Is the maintenance mode enabled in your CI build for these tests? (I assume you mean `ARANGODB_ENABLE_MAINTAINER_MODE`)

File: arangod/RestHandler/RestAdminStatisticsHandler.cpp [hunk 3]
```diff
@@ -56,6 +56,9 @@ RestStatus RestAdminStatisticsHandler::execute() {
   }
 
   if (_request->requestPath() == "/_admin/statistics") {
+    if (_request->value("dont-count-this-request") == "true") {
+      _statistics.SET_IGNORE();
+    }
```
Comment (jsteemann) [inline arangod/RestHandler/RestAdminStatisticsHandler.cpp:61]:
```suggestion
#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
    if (_request->value("dont-count-this-request") == "true") {
      _statistics.SET_IGNORE();
    }
#endif
```

File: tests/js/client/shell/api-statistics-spec.js [hunk 4]
```diff
@@ -49,44 +64,31 @@ function checkCommonStatisticsChanges(initialStats, finalStats) {
   for (const path of increaseByTwo) {
     expect(_.get(finalStats, path), path).to.be.equal(_.get(initialStats, path) + 2);
   }
-}
 
-describe('request statistics', function () {
+  // OPTIONS requests are not queue and therefore do not show up in queueTime
```
Comment (jsteemann) [inline tests/js/client/shell/api-statistics-spec.js:68]:
```suggestion
  // OPTIONS requests are not queued and therefore do not show up in queueTime
```

File: arangod/GeneralServer/VstCommTask.cpp [hunk 5]
```diff
@@ -252,6 +252,13 @@ bool VstCommTask<T>::processMessage(velocypack::Buffer<uint8_t> buffer,
...
+      const auto origin = req->header(StaticStrings::Origin);
```
Comment (jsteemann) [inline arangod/GeneralServer/VstCommTask.cpp:257]:
```suggestion
      std::string const& origin = req->header(StaticStrings::Origin);
```

File: arangod/GeneralServer/VstCommTask.cpp (hunk 1 above)
Comment (mpoeter) [reply arangod/GeneralServer/VstCommTask.cpp:256]: I was told that VST should behave as close as possible to HTTP. For OPTIONS requests that was not the case so far. What would you suggest instead?

Comment (graetzer) [reply arangod/GeneralServer/VstCommTask.cpp:256]: not handling OPTIONS requests perhaps? https://livebook.manning.com/book/cors-in-action/chapter-4/

Comment (graetzer) [reply arangod/GeneralServer/VstCommTask.cpp:256]: it only makes sense in the context of a webbrowser

Comment (jsteemann) [reply arangod/GeneralServer/VstCommTask.cpp:256]: Is there any harm in supporting this in VST, even though VST clients will make no use of it?
That way the differences in behavior would indeed be kept as small as possible. And IMHO it is unlikely that supporting OPTIONS requests in the server-side VST handler breaks anything.

Comment (mpoeter) [reply arangod/GeneralServer/VstCommTask.cpp:256]: I agree, that it doesn't really make sense for VST. Until now the integration tests have issued an OPTIONS request as indicator for when the statistics updates have been processed - that is why I stumbled over this difference in the first place.   
However, I have now found a simple solution to enforce a sync of the statistics update as part of the request (as @jsteemann has preferred in the first place), so from that perspective this change would be obsolete now. It is up to you - should I remove this change or keep it?

Comment (jsteemann) [reply arangod/GeneralServer/VstCommTask.cpp:256]: @mpoeter : ok, then let's remove OPTIONS handling from the VstCommTask for good.

**Comment by jsteemann** [comment]
Starting new Jenkins run after pulling in current devel:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10105/

**Comment by jsteemann** [comment]
Build failed on MSVC with narrowing warning:
``` 
  H2CommTask.cpp
C:\vm02-windows\oskar\work\ArangoDB\arangod\GeneralServer\H2CommTask.cpp(521,42): error C2220: the following warning is treated as an error [C:\vm02-windows\oskar\work\ArangoDB\build\arangod\arangoserver.vcxproj]
C:\vm02-windows\oskar\work\ArangoDB\arangod\GeneralServer\H2CommTask.cpp(501): message : while compiling class template member function 'void arangodb::rest::H2CommTask<arangodb::rest::SocketType::Tcp>::sendResponse(std::unique_ptr<arangodb::GeneralResponse,std::default_delete<arangodb::GeneralResponse>>,arangodb::RequestStatistics::Item)' [C:\vm02-windows\oskar\work\ArangoDB\build\arangod\arangoserver.vcxproj]
C:\vm02-windows\oskar\work\ArangoDB\arangod\GeneralServer\H2CommTask.cpp(804): message : see reference to class template instantiation 'arangodb::rest::H2CommTask<arangodb::rest::SocketType::Tcp>' being compiled [C:\vm02-windows\oskar\work\ArangoDB\build\arangod\arangoserver.vcxproj]
C:\vm02-windows\oskar\work\ArangoDB\arangod\GeneralServer\H2CommTask.cpp(521,42): warning C4244: 'argument': conversion from 'uint64_t' to 'int32_t', possible loss of data [C:\vm02-windows\oskar\work\ArangoDB\build\arangod\arangoserver.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\bin\HostX64\x64\clcache.py Finished. Exit code 2
```

**Comment by jsteemann** [comment]
Restarting test:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10108/

**Comment by jsteemann** [comment]
Thanks @mpoeter !

**Comment by mpoeter** [comment]
IMHO the fix itself feels a bit hackish - it is not immediately clear where the current statistics item resides (in the CommTask map, in the handler, in the stream?), but it was the simplest solution with the fewest code changes.

A cleaner solution would require more extensive design changes - something I might take on in the future.

**Comment by jsteemann** [comment]
@mpoeter : you are right, but the fix already helped a lot and is definitely a big improvement compared to what there was before.
Much appreciated!

**Pull request merged**